[iOS] The highlight selection box of multi-column extends to unselected content

cathiechen opened the pull request:
#### 59699ab765b97b10bea512f32eb14585d895ab30
<pre>
[iOS] The highlight selection box of multi-column extends to unselected content
<a href="https://bugs.webkit.org/show_bug.cgi?id=270787">https://bugs.webkit.org/show_bug.cgi?id=270787</a>

Reviewed by Wenson Hsieh.

This patch fixes two types of extending issues:
1. When selecting text from both inside and outside a multi-column layout, the column geometries merge with
the surrounding elements, causing the highlight box to extend across the entire width of the outer element.
This can make it appear as though the selection extends into an unselected column. To resolve this,
set m_separateFromPreviousLine for multi-column content. Identify multi-column content by checking
if the element has multiColumnFlow or if it is a descendant of RenderMultiColumnFlow.

2. Since the top of the second line in one column aligns with the bottom of the first line in the adjacent column,
selecting text that starts at the first line of a column and extends into the next column causes the geometries united.
To fix this, set m_separateFromPreviousLine for each column, ensuring that geometries with m_separateFromPreviousLine do not unite.

* LayoutTests/editing/selection/ios/select-content-from-multicolumn-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-multicolumn.html: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn-002-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn-002.html: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn.html: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn-002-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn-002.html: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn.html: Added.
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::currentNodeRequiresToSeparateLines): Return true if the current node is multi-collumn.
(WebCore::hasAncestorRequiresToSeparateLines): Return true if it is a descendant of a multi-collumn element.
(WebCore::RenderObject::collectSelectionGeometries): Do not unite the rect if the geometry is m_separateFromPreviousLine.
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::collectSelectionGeometries): Set m_separateFromPreviousLine true if linebox is isFirstAfterPageBreak.

Canonical link: <a href="https://commits.webkit.org/291765@main">https://commits.webkit.org/291765@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/024b4104a3dfeb09a362249ebe33c00feb05e322

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3294 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98977 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44496 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96019 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21985 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71705 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29051 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96971 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10279 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52044 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9964 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2535 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43812 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80220 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2617 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101017 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21021 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80714 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21273 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80859 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80075 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19931 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24622 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1982 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14178 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21005 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26183 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20692 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24152 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22433 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2635,14 +2635,25 @@ static void adjustTextDirectionForCoalescedGeometries(const SimpleRange& range,
+    if (currentRenderer->isOutOfFlowPositioned())
+        return true;
+
+    if (const auto* blockFlow = dynamicDowncast<const RenderBlockFlow>(*currentRenderer); blockFlow)
```
```suggestion
    if (CheckedPtr blockFlow = dynamicDowncast<const RenderBlockFlow>(*currentRenderer))
```

whsieh reviewed (approved):
LGTM with minor nit.

cathiechen committed:
[iOS] The highlight selection box of multi-column extends to unselected content
https://bugs.webkit.org/show_bug.cgi?id=270787

Reviewed by Wenson Hsieh.

This patch fixes two types of extending issues:
1. When selecting text from both inside and outside a multi-column layout, the column geometries merge with
the surrounding elements, causing the highlight box to extend across the entire width of the outer element.
This can make it appear as though the selection extends into an unselected column. To resolve this,
set m_separateFromPreviousLine for multi-column content. Identify multi-column content by checking
if the element has multiColumnFlow or if it is a descendant of RenderMultiColumnFlow.

2. Since the top of the second line in one column aligns with the bottom of the first line in the adjacent column,
selecting text that starts at the first line of a column and extends into the next column causes the geometries united.
To fix this, set m_separateFromPreviousLine for each column, ensuring that geometries with m_separateFromPreviousLine do not unite.

* LayoutTests/editing/selection/ios/select-content-from-multicolumn-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-from-multicolumn.html: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn-002-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn-002.html: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-inside-and-outside-multicolumn.html: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn-002-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn-002.html: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn-expected.txt: Added.
* LayoutTests/editing/selection/ios/select-content-outside-and-inside-multicolumn.html: Added.
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::currentNodeRequiresToSeparateLines): Return true if the current node is multi-collumn.
(WebCore::hasAncestorRequiresToSeparateLines): Return true if it is a descendant of a multi-collumn element.
(WebCore::RenderObject::collectSelectionGeometries): Do not unite the rect if the geometry is m_separateFromPreviousLine.
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::collectSelectionGeometries): Set m_separateFromPreviousLine true if linebox is isFirstAfterPageBreak.

Canonical link: https://commits.webkit.org/291765@main
